### ISSUE TITLE
Provide a disableSwipe to PageView

### ISSUE THREAD

**Issue by feinstein** [flutter/flutter#37510]
It would be nice to have a simple boolean inside `PageView` (or inside the `PageController`?) to disable the swipe. This way only the `PageController` could control which page we are showing.

This is useful for onboarding forms, where we want to block the user from going forward if there any validation erros on the data he provided (empty email field etc.), and advance the user programatically when he types something right, or we enable a Next button.

**Comment by S1r-Lanzelot** [comment]
Was looking for same thing, ended up using https://stackoverflow.com/questions/48043908/pageview-disable-the-default-scrolling-and-replace-it-with-tap-event?rq=1

**Comment by AmbrocioIsaias2808** [comment]
Yeah well i have an issue with this and a function similiar will be cool.
I'm developing an app, and i have a pageview that loads in background the different widgets in parallel with the initial page widget. The idea is that all the information loads and it´s available when the user swipes to the other pages of the pageview.

An now i have a problem, inside the main pageview i have a widget that uses another pageview to swipe between pages, it´s a pdfviewer, and the pdfviewer doesn´t reacts to gestures because it automaticly swipes the main pageview, i tried changing the swipe direction of the pdfviewer-pageview to vertical swiping but now it barely respond and when it does the widgets freezes the app.

Any ideas?

**Comment by feinstein** [comment]
1 - Questions should be asked on Stackoverflow.com

2 - I think this sounds like a case for `NestedScrollView` or something similar.

**Comment by Albert221** [comment]
I think this issue should be closed.

**Comment by feinstein** [comment]
Yes, I know, but I am asking for a simpler API to achieve the same thing.

**Comment by jigarpatel17** [comment]
Yes, to have api to enable/disable scrolling inside PageController will be very helpful. Thanks.

**Comment by brunotacca** [comment]
I vote for this... I want the ability to disable and enable at my needs (e.g. agreeing with a ToS).
It seems like I will have to go for one of the workarounds mentioned.

**Comment by brunotacca** [comment]
btw... I solved my problem by managing the **itemCount** variable.
Using mobx and having a observable currentIndex, I can wrap the **PageView.builder** in an **Observer** and then use an expression in the itemCount parameter.

```
Observer(
  builder: (_) => PageView.builder(
    itemCount: homePageStore.pageViewItemCountManaged,
    ...
  ),
)
```
then at your MobX store.
```
  @computed 
  int get pageViewItemCountManaged {
    if(canGoNextPage && currentPageIndex!=lastPage) {
      return currentPageIndex+2;
    } else {
      return currentPageIndex+1;
    }
  }
```

with this, I can control the observable canGoNextPage whenever I want.
hope it helps, it's way simpler than implementing a custom physics.

**Comment by jacob-hutchings** [comment]
I'm trying to disable swiping after `Draggable.onDragStarted` is called, but calling `setState` in `onDragStarted` to update `physics: NeverScrollableScrollPhysics()` is breaking `Draggable`. My `childWhenDragging: Container()` reverts to `child` after setState is called.

**Comment by feinstein** [comment]
Open a new issue with your case, as this might be more related to `Draggable`.

**Comment by ubbop42** [comment]
I did this. 
```      physics: viewModel.isLoading
          ? NeverScrollableScrollPhysics()
          : ScrollPhysics(),```

**Comment by glowinthedark** [comment]
This is a valid issue that needs to be addressed as the current API does not allow to dynamically enable/disable scrolling: `PageView.physics` is final and cannot be re-assigned dynamically once the object is created.

Also, there doesn't seem to be straightforward way to create a custom `ScrollPhysics` implementation that would allow to manage scrolling dynamically.

**Comment by feinstein** [comment]
Flutter is a reactive framework, if you want to change any parameters, then just rebuild the widget with the new value you want.

**Comment by glowinthedark** [comment]
@feinstein: do you by chance have a minimal example at hand which would show how to do it specifically for a `PageView`?

If I understand you correctly, you mean that `StatefulWidget` instance is unchanged while it is the `State<MyWidgetModel>` that has to be recreated whenever the state changes. How would that look in code? It would be very helpful for the whole community to understand this concept since it looks like many devs migrating from native android (including me) are expecting to do it the java-way which does not seem to be right approach in this case.

**Comment by feinstein** [comment]
I am at my phone right now so I will try some pseudo code/memory, correct it at your IDE, but it should look like this:

```dart
class MyWidget extends StatefulWidget {
... 
} 

class MyWidgetState extends State<MyWidget> {
  bool canScroll = true;

  Widget build(BuildContext context) {
    return Scaffold (
      body: PageView (
        physics: canScroll 
          ? NeverScrollableScrollPhysics()
           : ScrollPhysics(),
          ... 
       ),
       floatingActionButton: FloatingActionButton (
          onPressed: () {
             setState(() { can Scroll = false; });
         }, 
       ), 
   );
  } 
... 
} 
```

In any case Stackoverlow is the correct place to discuss this things, if you need further help, open a question there and send me the link.

**Comment by glowinthedark** [comment]
@feinstein: Thanks for the example! Much appreciated. I can confirm that `setState()` works as expected.

To everyone else who still considers this to be a bug: it is the way the flutter architecture is supposed to work — widgets are supposed to be **re-created** rather than **mutated** (which is probably what most android native java/kotlin devs are used to).

_Mandatory reading for understanding why **_this is not a defect_**:_

### Introduction to declarative UI
https://flutter.dev/docs/get-started/flutter-for/declarative

### State Management
https://flutter.dev/docs/development/data-and-backend/state-mgmt

### Simple app state management
https://flutter.dev/docs/development/data-and-backend/state-mgmt/simple

For my specific use case I needed to trigger a state change on a parent widget (i.e. programatically toggle the scrolling of a `PageView`) from a child widget which was possible by passing around a [callback](https://stackoverflow.com/questions/52082981/set-state-of-parent-widget) from the parent widget.

For more complex use cases which require recreating one widget based on changes in another widget instead of passing around a `Function` that wraps `setState()` you might consider using a [ChangeNotifier+ChangeNotifierProvider](https://flutter.dev/docs/development/data-and-backend/state-mgmt/simple#changenotifier)

**Comment by feinstein** [comment]
@glowinthedark as I said [here](https://github.com/flutter/flutter/issues/37510#issuecomment-612664222), I was looking for a simpler way to configure the scroll, it's not a bug report, it's a feature request.

**Comment by glowinthedark** [comment]
[ubbops42's answer](https://github.com/flutter/flutter/issues/37510#issuecomment-729207608) is as simple as it gets, you just need to wrap that call in `setState()` and you are done. [feinstein's example ](https://github.com/flutter/flutter/issues/37510#issuecomment-737884158) pretty much sums up the approach on doing it dynamically.

**Comment by SemicolonCG** [comment]
put NeverScrollableScrollPhysics() to physics and add on tap property to pageview to manually switch pages

**Comment by feinstein** [comment]
Please read the post before posting.

**Comment by netgfx** [comment]
I would also add disable forward and/or backward scrolling. If we are requesting an API to control the scrolling might as well try to handle the cases where we only want to disable forward or only backward scrolling.

**Comment by AimlessStudios** [comment]
None of the listed solutions are viable, as the pageview physics cant be changed without redrawing the entire pageview thus loosing state of all pages and the pageview itself.

I have a pageview each with a stack of card. The top card can be double tapped to 'lock the pageview' and enable 'Tinder' style swiping of the cards on that page, without engaging the scroll mechanics of the pageview. 

Thus this would be an EXTREMELY useful feature, as I have needed this functionality for the last 4 projects I have worked on, not including this one. I have spent over two months looking for a solution to this.

**Comment by felipe-xavier** [comment]
@AimlessStudios That is exactly my problem and the code you pasted should work? Is PageLockService your own class?

@glowinthedark I understand what you point out and I am using riverpod for updating the state in a child widget to know when I should pass `NeverScrollableScrollPhysics` or not similar to the logic @feinstein coded. But the thing is that this approach is causing my child widget to rebuild recursively, bcs this is making the PageView rebuild to update the physics param. So I really don't know what to do here. 

```
final _navBar = useProvider(bottomNavigationBarProvider);
return PageView(
      physics: !_navBar ? const NeverScrollableScrollPhysics() : null,
      controller: pageController,
      onPageChanged: onPageChanged,
      children: [
        Beamer(
          key: const Key('feed-tab'),
          routerDelegate: BeamerDelegate(locationBuilder: (state) => FeedLocation(state)),
        ), 
      ]  
```
So when `FeedLocation` loads, it updates `_navBar` in an attempt for disabling the scroll behavior. But as I mentioned, this causes the parent to rebuild and `FeedLocation` to build again and then the recursive state.. 

The idea was to be able to go to `FeedLocation`, disable the scroll, then when go back, enable it again, but I don't see a solution for that.

**Comment by felipe-xavier** [comment]
StackOverflow question:
https://pt.stackoverflow.com/questions/522829/flutter-how-to-disable-vertical-swipe-in-pageview

**Comment by Jack74r** [comment]
I agree, it's can be usefull and simple for simple form multi pages. 
Controller handle the pagination and verification between pages.
